internal/bminventory/inventory.go

```diff
@@ -325,6 +325,55 @@ func (b *bareMetalInventory) setPrimaryIPStack(cluster *common.Cluster) error {
 	return nil
 }
 
+// updatePrimaryIPStack handles primary IP stack validation and calculation for network changes
+// Returns true if PrimaryIPStack was updated, false otherwise
+func (b *bareMetalInventory) updatePrimaryIPStack(params installer.V2UpdateClusterParams, cluster *common.Cluster) (bool, error) {
+	// Only set primary IP stack if any networks or VIPs were actually updated
+	primaryIPStackNeedsUpdate := params.ClusterUpdateParams.ClusterNetworks != nil ||
+		params.ClusterUpdateParams.ServiceNetworks != nil ||
+		params.ClusterUpdateParams.MachineNetworks != nil ||
+		params.ClusterUpdateParams.APIVips != nil ||
+		params.ClusterUpdateParams.IngressVips != nil
+
+	if !primaryIPStackNeedsUpdate {
+		return false, nil
+	}
+
+	if (params.ClusterUpdateParams.ClusterNetworks != nil &&
+		params.ClusterUpdateParams.ServiceNetworks != nil &&
+		params.ClusterUpdateParams.MachineNetworks != nil &&
+		params.ClusterUpdateParams.APIVips != nil &&
+		params.ClusterUpdateParams.IngressVips != nil) || cluster.PrimaryIPStack == nil {
+		// Recalculate from scratch for full updates or new clusters
+		err := b.setPrimaryIPStack(cluster)
+		if err != nil {
+			b.log.WithError(err).Errorf("cluster update failed: unable to set primary IP stack")
+			return false, common.NewApiError(http.StatusBadRequest, err)
+		}
+		return true, nil // PrimaryIPStack was updated
+	}
+
+	if cluster.PrimaryIPStack != nil {
```

Comment:
Correct me if I am wrong, `cluster.PrimaryIPStack` can't be `nil` here due to the check in line `346`